### ISSUE TITLE
Add getter to access BatchNorm's private field

### ISSUE THREAD

**Issue by jamjamjon** [huggingface/candle#1149]
BatchNorm needs to offer getter to access its private field(running_var, running_mean, weight_and_bias, ...) in order to fuse Conv2d and BN. This will speedup inference a little bit(yolov8s, ~6ms).

**Comment by LaurentMazare** [comment]
Not sure to understand what you mean, could you give more context and detail a bit the benchmarks that you ran and what you measured?

**Comment by jamjamjon** [comment]
I need to fuse adjacent Conv2d()  and BatchNorm2d() layers together in order to speed-up inference time. It is usually achieved by eliminating the batch norm layer entirely and updating the weight and bias of the preceding convolution.  So I need to access the parameters (like running_var, running_mean, weight and bias) in BatchNorm.



### Here is the PyTorch code 
```
def fuse_conv_and_bn(conv, bn):
    fusedconv = nn.Conv2d(conv.in_channels,
                          conv.out_channels,
                          kernel_size=conv.kernel_size,
                          stride=conv.stride,
                          padding=conv.padding,
                          groups=conv.groups,
                          bias=True).requires_grad_(False).to(conv.weight.device)

    # Prepare filters
    w_conv = conv.weight.clone().view(conv.out_channels, -1)
    w_bn = torch.diag(bn.weight.div(torch.sqrt(bn.eps + bn.running_var)))
    fusedconv.weight.copy_(torch.mm(w_bn, w_conv).view(fusedconv.weight.shape))

    # Prepare spatial bias
    b_conv = torch.zeros(conv.weight.size(0), device=conv.weight.device) if conv.bias is None else conv.bias
    b_bn = bn.bias - bn.weight.mul(bn.running_mean).div(torch.sqrt(bn.running_var + bn.eps))
    fusedconv.bias.copy_(torch.mm(w_bn, b_conv.reshape(-1, 1)).reshape(-1) + b_bn)

    return fusedconv
```

### Rust code will be like
```

    fn fuse(&self, conv: &Conv2d, bn: &BatchNorm) -> Result<Conv2d> {
        let mut w_conv = conv.weight().clone();
        let (mut w_bn, mut b_bn) = bn.weight_and_bias.clone().unwrap();

        // fused conv weight ==> w_conv * (w_bn / sqrt(running_var + eps))
        let std_ = (bn.running_var.clone() + bn.eps)?.sqrt()?;
        let std_ = w_bn.div(&std_)?;
        let weight = w_conv.broadcast_mul(&(std_.reshape((bn.num_features, 1, 1, 1))?))?;

        // fused conv bias ==> bn_bias - w_bn * running_mean / sqrt(running_var + eps)
        let bias = b_bn.sub(&(std_.mul(&bn.running_mean)?))?;

        // create new conv with bias
        Ok(Conv2d::new(weight, Some(bias), conv.config().clone()))
    }
```


For now, all fields are private and offer no methods to access them.
```
#[derive(Clone, Debug)]
pub struct BatchNorm {
    running_mean: Tensor,
    running_var: Tensor,
    weight_and_bias: Option<(Tensor, Tensor)>,
    remove_mean: bool,
    eps: f64,
    num_features: usize,
}
```
https://github.com/huggingface/candle/blob/c698e176190eca52ff754e10d05becc972c75a5c/candle-nn/src/batch_norm.rs#L41C1-L50C1

So some getter functions of BatchNorm should be available to retrieve the private fields.

**Comment by LaurentMazare** [comment]
Sounds like a compelling use case, thanks for the details. I've just merged #1176 to expose these.
Also if you want to have your fuse function available in `candle_nn`, feel free to make a PR, this seem like a generally helpful thing to have.

**Comment by jamjamjon** [comment]
I will make a PR later on today.
